### PR TITLE
Handle the --namespace flag

### PR DESCRIPTION
**Pull request by mortent**
The `--namespace` flag is currently not used by any of the commands, but it is available as a flag. This PR updates the commands to actually use both the namespace provided in the context as well as the namespace provided with the flag. It behaves as follows:

For `kapply init`:
 * If the user has not provided a namespace for the inventory with the `--inventory-namespace` flag and none of the resources in the package has the namespace set, it will fall back to use the namespace provided with the `--namespace` flag and then to the namespace provided in the context.

For `kapply apply/preview`:
 * If a resource does not have the namespace set and the resource has namespace scope, it will fall back to use the namespace provided with the `--namespace` flag and then to the namespace provided in the context. The inventory template is ignored in this process, so this will never affect the namespace for the inventory
 * If the namespace is provided with the `--namespace` flag, it is considered an error if any of the resources (except the inventory) is in a different namespace. This is aligned with the behavior in kubectl. 

Fixes: https://github.com/GoogleContainerTools/kpt/issues/975, https://github.com/GoogleContainerTools/kpt/issues/826

### CODE REVIEW

**Comment by wpwoodjr** [comment]
>If the user has not provided a namespace for the inventory with the --inventory-namespace flag and none of the resources in the package has the namespace set, it will fall back to use the namespace provided with the --namespace flag and then to the namespace provided in the context.

What happens if the user has not provided a namespace for the inventory with the --inventory-namespace flag and a resource **does** have the namespace set?

File: pkg/config/initoptions.go
Comment (seans3) [inline pkg/config/initoptions.go:150]: We probably want to add this new parameter (and what it means) to the function documentation.

File: cmd/apply/cmdapply.go
Comment (seans3) [inline cmd/apply/cmdapply.go:126]: Are we using the `EnforceNamespace` field anywhere?

File: cmd/preview/cmdpreview.go
Comment (seans3) [inline cmd/preview/cmdpreview.go:116]: Maybe a comment here for those not familiar with namespace within the kubeconfig file.

File: cmd/apply/cmdapply.go
Comment (seans3) [inline cmd/apply/cmdapply.go:117]: Same about a comment here.

**Comment by wpwoodjr** [comment]
I'm not clear why a separate `--inventory-namespace` flag is needed.

For `kpt apply init` why can't the `--namespace` flag set the inventory namespace?  That's what I expected initially until I saw the `--inventory-namespace` flag.

**Comment by mortent** [comment]
@wpwoodjr I did consider using the `--namespace` flag directly, but it is a flag that is used across all `kpt live` commands as well as in kubectl. The way it is handled now, it should behave in a way that is consistent with how it is used elsewhere. The way the namespace flag interacts with the namespace set in the context also makes it hard to use it instead of the `--inventory-namespace` flag.

So with this change, the `--inventory-template` flag is a way to force the namespace to a specific value, while the `--namespace` flag and the namespace set on the context provides a way to change the default when none of the resources have the namespace set.

File: cmd/apply/cmdapply.go
Comment (mortent) [reply cmd/apply/cmdapply.go:126]: The goal is to use it the same way here as it is used in kubectl. So the if the value of the namespace is coming from the context, the value of `EnforceNamespace` will be `false`, while if the namespace is set with the `--namespace` flag, the `EnforceNamespace` value will be `true`.
The value here is used in the `setNamespaces` function.

**Comment by wpwoodjr** [comment]
> @wpwoodjr I did consider using the `--namespace` flag directly, but it is a flag that is used across all `kpt live` commands as well as in kubectl. The way it is handled now, it should behave in a way that is consistent with how it is used elsewhere. The way the namespace flag interacts with the namespace set in the context also makes it hard to use it instead of the `--inventory-namespace` flag.
> 
> So with this change, the `--inventory-template` flag is a way to force the namespace to a specific value, while the `--namespace` flag and the namespace set on the context provides a way to change the default when none of the resources have the namespace set.

I think its confusing.  Why can't it just ignore the namespace if set on other resources. `kpt live init .` could first use the `--namespace` flag, if that's not present fall back to the context, if that's not present fall back to `default` (or even an error)?

**Comment by mortent** [comment]
So an alternative solution here is the following:
 * If the namespace is provided with the `--namespace` flag, that namespace will be used for the inventory no matter which namespaces are in the resources.
 * If the `--namespace` flag is not set and all resources have a namespace set and it is the same namespace, that will also be used for the inventory.
 * In all other cases, the namespace from the context will be used. This is always specified, and in most cases it will be `default`.

If we adopted this solution, we should also change the name of the `--inventory-name` flag to `--name` to be consistent. 

@wpwoodjr Do you think this approach is less confusing?

**Comment by wpwoodjr** [comment]
@mortent yes I think it is less confusing, thanks!

**Comment by seans3** [comment]
Some more context on the `--namespace` flag. As mentioned, it is used by all the `kpt live` commands that speak to an API Server (`kpt live apply`, `kpt live preview`). It is defined in the same place as all the `kubeconfig` flags that are used to communicate with an API Server. Here is the set of `kubeconfig` flags (including Namespace): `k8s.io/kubernetes/staging/src/k8s.io/cli-runtime/pkg/genericclioptions/config_flags.go`
```
/ ConfigFlags composes the set of values necessary                                                                                                                                                           
// for obtaining a REST client config                                                                                                                                                                         
type ConfigFlags struct {
    CacheDir   *string
    KubeConfig *string

    // config flags                                                                                                                                                                                           
    ClusterName      *string
    AuthInfoName     *string
    Context          *string
    Namespace        *string
    APIServer        *string
    TLSServerName    *string
    Insecure         *bool
    CertFile         *string
    KeyFile          *string
    CAFile           *string
    BearerToken      *string
    Impersonate      *string
    ImpersonateGroup *[]string
    Username         *string
    Password         *string
    Timeout          *string

    clientConfig clientcmd.ClientConfig
    lock         sync.Mutex
    // If set to true, will use persistent client config and                                                                                                                                                  
    // propagate the config to the places that need it, rather than                                                                                                                                           
    // loading the config multiple times                                                                                                                                                                      
    usePersistentConfig bool
}
```

The reason the `--namespace` flag can be re-used (to specify the inventory namespace) in the `kpt live init` command because we do not contact the API server for this command. If the `kpt live init` command ever needs to interact with the cluster, then this `--namespace` flag reuse will not work.

**Comment by seans3** [comment]
I'll LGTM when the merge conflict is resolved.

**Comment by mortent** [comment]
@seans3 This is updated
/hold cancel

**Comment by wpwoodjr** [comment]
> The reason the `--namespace` flag can be re-used (to specify the inventory namespace) in the `kpt live init` command because we do not contact the API server for this command. If the `kpt live init` command ever needs to interact with the cluster, then this `--namespace` flag reuse will not work.

Its hard to comment on a future, unspecified need; but why would the namespace need to be different if you were contacting the server in that instance?

**Review by seans3 — APPROVED** [review]
/lgtm

**Pull request merged**
